teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
 a managed Tekton, users can be prevented access to the `tekton-pipelines`
 namespaces, where the feature flag config map is hosted.
 
+## Proposal
+
+We will tackle separately features that are controlled via new fields in the API
+and features that are only behavioral. (The alternative
+[one feature flag per new alpha field](#one-feature-flag-per-new-alpha-field) describes
+an alternative where we use the same approach for both.)
+
+In both cases we want to control access with
+[a feature flag](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior).
+
+* For new API fields, will will gate access to all of them behind two feature flags
+  called `enable-alpha-api-fields` and `enable-beta-api-fields`.
+  ([Design details: new api fields](#new-api-fields).)
+* For changes in behavior which are not configurable via API fields in our CRDs,
+  each feature will be gated by its own unique feature flag.
+  ([Design details: changes in behavior](#changes-in-behavior).)
+
+
+### Risks and Mitigations
+
+See the pros and cons sections in the [design details](#design-details)
+below.
+
+## Design Details
+
+### New API fields
+
+We gate access to all new API fields with two new
+[feature flags](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior) called `enable-alpha-api-fields` and `enable-beta-api-fields`
+which will default to false.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  enable-alpha-api-fields: "true"
+  enable-beta-api-fields: "true"
+```
+
+This allows administrators to opt into allowing their users to use alpha and beta fields.
+
+_Note that we will only need the `enable-beta-api-fields` flag once we have v1 types in Tekton,
+at the moment the highest version we have is beta, so we will initially only need
+`enable-alpha-api-fields`._
+
+Anyone submitting any Tekton CRD (e.g. Pipelines, Tasks, Runs, PipelineRuns, TaskRuns) when these are
+true will be able to use fields which are considered "alpha" and/or "beta" (depending on which fields
+are set). The API fields enabled by these fields may be changed according to our
+[API compatibility policy](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md)
+meaning that they could be dropped or changed with at least one release of warning. (This will be true
+even when the fields are added to types which are considered `beta` or `ga`/`v1`).
+
+Since the API fields will exist in the CRD with or without the flags, this will be enforced at runtime
+by the webhook admission controller: if the flag is not enabled and a user tries to use an
```

Comment:
> Is that right?

Based on my understanding, yes, that is correct. An object creation will fail if `alpha` field is disabled but specified in the resource definition.

> 3. Cluster admin disables alpha features

Should we run any kind of validation here, saying you have so many tasks created with `alpha` features which will no longer be usable after you disable `alpha` features?

> Are the Task's alpha versioned fields validated and rejected at this point?

Yes, I think so, we will need extra validation during the task consumption time in addition to task creation time.

Reply:
What about this scenario:

0. Cluster has alpha features enabled
1. User writes Task using alpha fields
2. User applies Task with `kubectl`
3. **Cluster admin updates Pipelines where the alpha feature was upgraded to beta**
4. User applies beta TaskRun with `taskRef:` to alpha Task

This would run into a failure as well since `beta` feature is used but is not enabled. But would this be allowed if `beta` was enabled in step 0.?

Also, thinking more on this, is this kind of implementation possible under single API? 

| Features | Alpha | Beta | GA |
| --- | --- | --- | --- |
| A | x | | |
| B | x | | |
| C |  | x | |

If so, for this configuration, any cluster will need `alpha` and `beta` both enabled.

And no changes needed when the feature `A` upgrades to `beta`:

| Features | Alpha | Beta | GA |
| --- | --- | --- | --- |
| A | x | | |
| B |  | x  | |
| C |  | x | |

Right?

Reply:
Very good question @sbwsg and @pritidesai. My feeling is that we should fail in that case (fields were enable and are now disabled). The question is *when* — we probably cannot fail at the *creation* phase (because the `Task` is already there) but we can at the execution time (when we fetch the task spec, we can validate wether those field are allowed or not in the current configuration).

@pritidesai not sure I followed. Are we assuming the cluster admin had enabled beta but not alpha ? If that's true, on *upgrade*, the alpha features that weren't allowed before are now *magically* allowed (because they can be used per the configuration). 

I am not sure I understand the tables 😅. Is there any relation between A, B and C (C depends on A or something) ? because if there isn't, I don't see any problem. A user that only enable "ga api feature" won't be able to use A, B or C, a user that enable "beta api feature" will be able to use C in the first table, and B and C in the 2nd table.

Reply:
@sbwsg I'm adding an example of what would happen if you remove an alpha or beta field and trying to capture the case you highlighted where something in the cluster becomes invalid

@pritidesai that's a great point - i think that @afrittoli 's suggested alternative syntax will make this a lot easier! (i.e. just one flag, where setting "alpha" will imply enabling "beta" as well) (let me know if im not understanding tho!)